ekm/remote_key_manager.go

```diff
@@ -0,0 +1,414 @@
+	logger          *zap.Logger
+	remoteSigner    RemoteSigner
+	consensusClient ConsensusClient
+	getOperatorId   func() spectypes.OperatorID
+	operatorPubKey  keys.OperatorPublicKey
+	SlashingProtector
+}
+
+type RemoteSigner interface {
+	AddValidators(ctx context.Context, shares ...ssvsigner.ClientShareKeys) ([]web3signer.Status, error)
+	RemoveValidators(ctx context.Context, sharePubKeys ...[]byte) ([]web3signer.Status, error)
+	Sign(ctx context.Context, sharePubKey []byte, payload web3signer.SignRequest) ([]byte, error)
+	OperatorIdentity(ctx context.Context) (string, error)
+	OperatorSign(ctx context.Context, payload []byte) ([]byte, error)
+}
+
+type ConsensusClient interface {
+	CurrentFork(ctx context.Context) (*phase0.Fork, error)
+	Genesis(ctx context.Context) (*eth2apiv1.Genesis, error)
+}
+
+func NewRemoteKeyManager(
+	logger *zap.Logger,
+	remoteSigner RemoteSigner,
+	consensusClient ConsensusClient,
+	db basedb.Database,
+	networkConfig networkconfig.NetworkConfig,
+	getOperatorId func() spectypes.OperatorID,
+) (*RemoteKeyManager, error) {
+	signerStore := NewSignerStorage(db, networkConfig.Beacon, logger)
+	protection := slashingprotection.NewNormalProtection(signerStore)
+
+	operatorPubKeyString, err := remoteSigner.OperatorIdentity(context.Background()) // TODO: use context
+	if err != nil {
+		return nil, fmt.Errorf("get operator identity: %w", err)
+	}
+
+	operatorPubKey, err := keys.PublicKeyFromString(operatorPubKeyString)
+	if err != nil {
+		return nil, fmt.Errorf("extract operator public key: %w", err)
+	}
+
+	return &RemoteKeyManager{
+		logger:            logger,
+		remoteSigner:      remoteSigner,
+		consensusClient:   consensusClient,
+		SlashingProtector: NewSlashingProtector(logger, signerStore, protection),
+		getOperatorId:     getOperatorId,
+		operatorPubKey:    operatorPubKey,
+	}, nil
+}
+
+func (km *RemoteKeyManager) AddShare(encryptedSharePrivKey, sharePubKey []byte) error {
+	shareKeys := ssvsigner.ClientShareKeys{
+		EncryptedPrivKey: encryptedSharePrivKey,
+		PublicKey:        sharePubKey,
+	}
+
+	statuses, err := km.remoteSigner.AddValidators(context.Background(), shareKeys) // TODO: pass context from outside
+	if err != nil {
+		return fmt.Errorf("add validator: %w", err)
+	}
+
+	if len(statuses) != 1 {
+		return fmt.Errorf("bug: expected 1 status, got %d", len(statuses))
+	}
+
+	if statuses[0] != web3signer.StatusImported {
+		return fmt.Errorf("unexpected status %s", statuses[0])
+	}
+
+	if err := km.BumpSlashingProtection(sharePubKey); err != nil {
+		return fmt.Errorf("could not bump slashing protection: %w", err)
+	}
+
+	return nil
+}
+
+func (km *RemoteKeyManager) RemoveShare(pubKey []byte) error {
+	statuses, err := km.remoteSigner.RemoveValidators(context.Background(), pubKey) // TODO: use context
+	if err != nil {
+		return fmt.Errorf("remove validator: %w", err)
+	}
+
+	if len(statuses) != 1 {
+		return fmt.Errorf("bug: expected 1 status, got %d", len(statuses))
+	}
```

Comment:
shouldn't this check be done in the client methods if they're expected to return len(statuses) == len(pubkeys) always?

Reply:
it's done in the client but it seems to confuse reviewers so I added it here too